buffer/src/main/java/io/netty/buffer/api/adaptor/ByteBufBuffer.java

```diff
@@ -0,0 +1,1312 @@
+     *
+     * @param byteBuf The {@link ByteBuf} instance to wrap.
+     * @return A new {@link Buffer} instance, backed by the given {@link ByteBuf}.
+     */
+    public static Buffer wrap(ByteBuf byteBuf) {
+        Buffer buffer = ByteBufAdaptor.extract(byteBuf);
+        if (buffer == null) {
+            buffer = new ByteBufBuffer(new ByteBufAllocatorControl(byteBuf), byteBuf, ByteBufDrop.INSTANCE);
+        }
+        return buffer;
+    }
+
+    /**
+     * Unwrap this {@link Buffer} instance by returning the inner {@link ByteBuf} instance.
+     * <p>
+     * This method destroys this {@link Buffer} instance, as if {@link #close()} is called upon it,
+     * but without harm to the returned {@link ByteBuf} instance.
+     *
+     * @return The inner {@link ByteBuf} instance.
+     */
+    public ByteBuf unwrapAndClose() {
+        try (ByteBufBuffer ignore = this) {
+            return delegate.retain();
+        }
+    }
+
+    @Override
+    public String toString() {
+        return "Buffer[roff:" + delegate.readerIndex() + ", woff:" + delegate.writerIndex() +
+               ", cap:" + delegate.capacity() + ']';
+    }
+
+    @Override
+    protected RuntimeException createResourceClosedException() {
+        return bufferIsClosed(this);
+    }
+
+    @Override
+    protected Owned<ByteBufBuffer> prepareSend() {
+        final ByteBuf delegate = this.delegate;
+        return new Owned<ByteBufBuffer>() {
+            @Override
+            public ByteBufBuffer transferOwnership(Drop<ByteBufBuffer> drop) {
+                return new ByteBufBuffer(control, delegate, drop);
+            }
+        };
+    }
+
+    @Override
+    protected void makeInaccessible() {
+        super.makeInaccessible();
+        delegate = (ByteBuf) CLOSED_BYTEBUF;
+    }
+
+    @Override
+    public int capacity() {
+        return delegate.capacity();
+    }
+
+    @Override
+    public int readerOffset() {
+        return delegate.readerIndex();
+    }
+
+    @Override
+    public void skipReadable(int delta) {
+        delegate.readerIndex(delegate.readerIndex() + delta);
+    }
+
+    @Override
+    public Buffer readerOffset(int offset) {
+        delegate.readerIndex(offset);
+        return this;
+    }
+
+    @Override
+    public int writerOffset() {
+        return delegate.writerIndex();
+    }
+
+    @Override
+    public void skipWritable(int delta) {
+        delegate.writerIndex(delegate.writerIndex() + delta);
+    }
+
+    @Override
+    public Buffer writerOffset(int offset) {
+        if (readOnly()) {
+            throw bufferIsReadOnly(this);
+        }
+        delegate.writerIndex(offset);
+        return this;
+    }
+
+    @Override
+    public int readableBytes() {
+        return delegate.readableBytes();
+    }
+
+    @Override
+    public int writableBytes() {
+        return delegate.capacity() - delegate.writerIndex();
```

Comment:
```suggestion
        return delegate.writableBytes();
```

Reply:
No, that breaks for read-only buffers. I did try it. 🤷‍♂️

Reply:
in what way ? Because it always return 0 ?

Reply:
Yes. In the Buffer interface, it's expected to just be based on the write offset and the capacity.

Reply:
I am a bit on the fence here... Generally speaking it seems odd that a `ReadOnly` buffer would return anything different then `0` for `writableBytes`.